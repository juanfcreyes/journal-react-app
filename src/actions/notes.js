import Swal from "sweetalert2";
import { db } from "../firebase/firebase-config";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotesDB } from "../helpers/loadNotesDB";
import { types } from "../types/types";

export const startNewNote = () => {
    return async (dispatch, getState) => {
        const { auth } = getState();
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }
        try {
            const doc = await db.collection(`${auth.uid}/journal/notes`).add(newNote);
            dispatch(setActiveNote(doc.id, newNote));
            dispatch(addNewNote({ id: doc.id, ...newNote}));
        } catch (error) {
            console.error(error);
        }
    
    }
};

export const startLoadNotes = () => {
    return async (dispatch, getState) => {
        const { auth } = getState();
        try {
            const notes = await loadNotesDB(auth.uid);
            dispatch(loadNotes(notes));
        } catch (error) {
            dispatch(loadNotes([]));
            console.log(error);
        }
    }
};

export const startSaveNote = (note) => {
    return async (dispatch, getState) => {
        const { auth } = getState();
        const noteFS = {...note};
        delete noteFS.id;
        try {
            await db.doc(`${auth.uid}/journal/notes/${note.id}`).update(noteFS);     
        } catch (error) {
            console.log('startSaveNote error', error)
        }
        dispatch(refreshNote(note.id, noteFS));
        dispatch(setActiveNote(note.id, noteFS))
        Swal.fire('Save', note.title, 'success');
    }
}

export const startUploading = (file) => {
    return async (dispatch, getState) => {
        const { active: note } = getState().notes;
        showUploadAlert();
        try {
            const fileUrl = await fileUpload(file);
            note.url = fileUrl;        
        } catch (error) {
            
        }
        dispatch(startSaveNote(note))
        Swal.close();
    }
}

export const startDeleting = (id) => {
    return async (dispatch, getState) => {
        try {
            const { auth } = getState();
            await db.doc(`${auth.uid}/journal/notes/${id}`).delete();
            Swal.fire('Deleted', 'Note deleted', 'info');
            dispatch(deleteNote(id));
        } catch (error) {
            console.error(error);
        }
      
    }
}

const showUploadAlert = () => {
    Swal.fire({
        title: 'Uploading...',
        text: 'Please wait',
        allowOutsideClick: false,
        willOpen: () => {
            Swal.showLoading()
        },
        showCloseButton: false,
        showCancelButton: false,
        showConfirmButton: false
    });
} 

export const addNewNote = (note) => ({
    type: types.notesAddnew,
    payload: {
        ...note
    }
});

export const setActiveNote = (id, note) => ({
    type: types.notesSetActive,
    payload: {
        id,
        ...note
    }
});

export const loadNotes = (notes) => ({
    type : types.notesLoad,
    payload: notes
});

export const refreshNote = (id, note) => ({
    type: types.notesUploaded,
    payload: {
        id,
        note: { id, ...note }
    }
});

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: {
        id
    }
});

export const notesLogout = () => ({
    type: types.notesCleaning,
});

