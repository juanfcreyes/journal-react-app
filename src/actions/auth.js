import { types } from "../types/types";
import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { finishLoading, startLoading } from "./ui";
import Swal from "sweetalert2";
import { notesLogout } from "./notes";

export const startLogin = (email, password) => {
    return async (dispatch) => {
        dispatch(startLoading());
        return firebase.auth().signInWithEmailAndPassword(email, password).then(({user}) => {
            dispatch(login(user.uid, user.displayName));
            dispatch(finishLoading());
        }).catch((error) => {
            Swal.fire('Error', error.message, 'error');
            dispatch(finishLoading());
            console.log('error', error);
        })
        
    }
}

export const startRegister = (email, password, name) => {
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email, password).then(async ({user}) => {
            await user.updateProfile({displayName: name})
            dispatch(login(user.uid, user.displayName));
        }).catch((error) => {
            Swal.fire('Error', error.message, 'error');
        });
    }
}

export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider).then(({ user }) => {
            dispatch(login(user.uid, user.displayName));
        });
    }
}

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();
        dispatch(logout());
        dispatch(notesLogout()); 
    }
}

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
});

export const logout = () => ({
    type: types.logout
});