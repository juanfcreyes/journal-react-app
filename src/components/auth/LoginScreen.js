import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import { startLogin, startGoogleLogin } from '../../actions/auth'
import { useForm } from '../../hooks/useForm'

export const LoginScreen = () => {
    
    const ui = useSelector(state => state.ui);
    const dispatch = useDispatch();
        
    const [formValues, handleInputChange] = useForm(
        {
            email:'',
            password: ''
        });

    const {email, password} = formValues;

    const handleLogin = (e) => {
        e.preventDefault(); 
        dispatch(startLogin(email, password));
    }

    const handleGoogleLogin = (e) => {
        e.preventDefault(); 
        dispatch(startGoogleLogin());
    }
    
    return (
        <>
            <h1 className='auth__title'> Login Screen</h1>
            <form onSubmit={handleLogin} className='animate__animated animate__fadeIn'>
                <input className='auth__input' autoComplete='off' type='text' placeholder='email' name='email' value={email} onChange={handleInputChange}/>
                <input className='auth__input' type='password' placeholder='password' name='password' value={password} onChange={handleInputChange}/>
                <button className='btn btn-primary btn-block mb-1 mt-1' type='submit' disabled={ui.loading}>Login</button>
                 <div className='auth__social-networks mb-1'>  
                    <p className='mb-1'>Login with social networks</p>
                    <div className="google-btn" onClick={handleGoogleLogin}>
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>
                <Link to='/auth/register' className='link'>
                    Create new account
                </Link>
            </form>
        </>
    )
}
