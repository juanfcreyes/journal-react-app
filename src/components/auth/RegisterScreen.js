import React from 'react'
import validator from 'validator';
import { useDispatch, useSelector } from 'react-redux'
import { useForm } from '../../hooks/useForm'
import { Link } from 'react-router-dom'
import { removeError, setError } from '../../actions/ui';
import { startRegister } from '../../actions/auth';

export const RegisterScreen = () => {
    
    const ui = useSelector(state => state.ui); 
    const dispatch = useDispatch();

    const [formValues, handleInputChange] = useForm(
        {
            name: '',
            email:'',
            password1: '',
            password2: ''
        });

    const {name, email, password1, password2} = formValues;

    const handleRegister = (e) => {
        e.preventDefault();
        if (isFormValid()) {
            dispatch(startRegister(email, password1, name));
        }
    }

    const isFormValid = () => {

        if (name.trim().length === 0) {
            dispatch(setError('Name is required'));
            return false;
        }

        if (!validator.isEmail(email)) {
            dispatch(setError('Email not valid'));
            return false;
        }

        if (password1 !== password2 || password1.length < 5) {
            dispatch(setError('The password must have more than 5 characters and its confirmation must be the same'));
            return false;
        } 

        dispatch(removeError());

        return true;
    }

    return (
        <>
            <h1 className='auth__title'> Register</h1>
            <form onSubmit={handleRegister}  className='animate__animated animate__fadeIn'>
                {
                    ui.msgError &&
                    <div className='auth__alert-error'>        
                        { ui.msgError }
                    </div>
                }
                <input className='auth__input' autoComplete='off' type='text' placeholder='name' name='name' onChange={handleInputChange}  value={name}/>
                <input className='auth__input' autoComplete='off' type='text' placeholder='email' name='email' onChange={handleInputChange} value={email}/>
                <input className='auth__input' type='password' placeholder='password' name='password1' onChange={handleInputChange} value={password1}/>
                <input className='auth__input' type='password' placeholder='confirm password' name='password2' onChange={handleInputChange} value={password2}/>
                <button className='btn btn-primary btn-block mb-5 mt-1' type='submit'>Register</button>
                <Link to='/auth/login' className='link'>
                   have you already registered?
                </Link>
            </form>
        </>
    )
}
