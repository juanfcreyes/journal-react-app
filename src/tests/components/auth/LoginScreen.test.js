import React from 'react'
import { mount } from "enzyme";
import { Provider } from "react-redux";
import { LoginScreen } from "../../../components/auth/LoginScreen";

import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk';
import { MemoryRouter } from "react-router-dom";
import { startGoogleLogin, startLogin } from '../../../actions/auth';

jest.mock('../../../actions/auth', () => ({
    startGoogleLogin: jest.fn(),
    startLogin: jest.fn()
}));


const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
    ui: {
        loading: false,
        msgError: null
    }
}

let store = mockStore(initState);
store.dispatch = jest.fn();

describe('Pruebas sobre LoginScreen', () => {
    
   const wrapper = mount(
    <Provider store={store}>
        <MemoryRouter>
            <LoginScreen></LoginScreen>
        </MemoryRouter>
    </Provider>
   );
   
   beforeEach(() => {
       store = mockStore(initState);
       jest.clearAllMocks();
    });

    test('Se debe renderizar correctamente  ', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de disparar la accion startGoogleLogin', () => {
        wrapper.find('.google-btn').prop('onClick')({
            preventDefault: () => {}
        });

        expect(startGoogleLogin).toHaveBeenCalled();
    });
    
    
    test('Debe de disparar la accion startLogin con los respectivos arguentos', () => {
        const email = 'juan.test@test.com';
        const password = 'test.123';
        wrapper.find('input[name="email"]').simulate('change', {
            target: {
                name: 'email',
                value: 'juan.test@test.com'
            }
        });
        wrapper.find('input[name="password"]').simulate('change', {
            target: {
                name: 'password',
                value: 'test.123'
            }
        });
        
        wrapper.find('form').prop('onSubmit')({
            preventDefault: () => {}
        });
        
        expect(startLogin).toHaveBeenCalledWith(email, password);
    });
    
});
