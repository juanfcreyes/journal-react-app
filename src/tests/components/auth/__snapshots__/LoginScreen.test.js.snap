// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre LoginScreen Se debe renderizar correctamente   1`] = `
Array [
  <h1
    className="auth__title"
  >
     Login Screen
  </h1>,
  <form
    className="animate__animated animate__fadeIn"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="email"
      type="text"
      value=""
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="password"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block mb-1 mt-1"
      disabled={false}
      type="submit"
    >
      Login
    </button>
    <div
      className="auth__social-networks mb-1"
    >
      <p
        className="mb-1"
      >
        Login with social networks
      </p>
      <div
        className="google-btn"
        onClick={[Function]}
      >
        <div
          className="google-icon-wrapper"
        >
          <img
            alt="google button"
            className="google-icon"
            src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
          />
        </div>
        <p
          className="btn-text"
        >
          <b>
            Sign in with google
          </b>
        </p>
      </div>
    </div>
    <a
      className="link"
      href="/auth/register"
      onClick={[Function]}
    >
      Create new account
    </a>
  </form>,
]
`;
